### PR TITLE
Add semantic segmentation output

### PR DESCRIPTION
**Pull request by DJayalath**
## Brief
Introduces semantic segmentation to the forest generator. When enabled, outputs to frame folder with frame--seg-<>.png similar to depth and colour. This addition is fully backwards compatible with previous yaml configuration files as the parameters added are optional.

## Why?
Semantic segmentation is useful to train computer vision applications e.g. imitation learning in simulation.

## What was changed?
- Add parameter to enable semantic segmentation output (similar to depth output)
- Add parameter to enable reasonable random frames (useful for unbiased data collection independent of time domain)
- Introduce segmentation colour to mesh properties
- Fragment shader will colour meshes accordingly if segmentation is enabled
- Certain features have a fixed colour in semantic segmentation e.g. sky
- Add two example yaml files (segmentation.yaml and segmentationRandom.yaml) for demonstration
- Update README to reflect changes

## How to use?
Add `segmentation: true` under output in configuration to enable segmentation outputs. Add
 ```
random:
   enabled: false
   endMax: 1000 # Number of random frames to collect
```
under input in configuration to collect data from random positions and reasonable angles if desired.

### Colour
image
### Segmentation
image

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -185,6 +187,10 @@ random:
    seed: -1 # A value of -1 will result in a seed derived from the system clock value
 
 input:
+  # Random data collection
+  random:
```
Comment (callumnewlands) [inline README.md]: As KPZ says probably best to rename this to be more descriptive, maybe something like "randomDataCollection" or "useRandomFrames" or "useRandomPositions" -- whichever you think fits best

File: README.md [hunk 2]
```diff
@@ -185,6 +187,10 @@ random:
...
+   enabled: false
+   endMax: 1000
```
Comment (callumnewlands) [inline README.md]: Add a comment here describing how endMax works

File: README.md [hunk 3]
```diff
@@ -211,6 +217,8 @@ output:
    colour: true
    # Output a depth image
    depth: false # If outputting to window the depth output will be displayed instead of colour if this is true, if rendering to images, both can be output
+   # Output a segmentation map
+   segmentation: true
```
Comment (callumnewlands) [inline README.md]: Should probably be false by default

File: src/main/java/App.java [hunk 4]
```diff
@@ -715,6 +717,30 @@ private void renderShadowMap() {
...
+
+				camera.setPosition(new Vector3f(rX, height, rZ));
+
+				camera.setYaw(rYaw);
+				camera.setPitch(rPitch);
+				camera.updateVectorsFromAngles();
+
+				if (frame >= parameters.input.random.endMax - 1) {
+					exit();
+				}
```
Comment (callumnewlands) [inline src/main/java/App.java]: Might be best to extract this into a separate method for readability but I'll leave that up to your discresion

File: src/main/java/segmentation/Colour.java [hunk 5]
```diff
@@ -0,0 +1,15 @@
+package segmentation;
+
+import org.joml.Vector3f;
+
+public class Colour {
+
```
Comment (callumnewlands) [inline src/main/java/segmentation/Colour.java:6]: Add a private (default) constructor to enforce static, singleton pattern (prevent instantiation of Colour class)

File: src/main/resources/segmentation.yaml [hunk 6]
```diff
@@ -0,0 +1,742 @@
...
+      numSamples: 100
+      # Density of samples along each light ray (higher = samples taken from a smaller area = shorter, brighter rays)
+      sampleDensity: 1.3
+      # Impact decay of each sample (lower = greater fall-off)
+      decay: 0.99 # range [0.0, 1.0)
+      # Scalar for brightness of each ray (higher = brighter)
+      exposure: 0.0015
+      # Brightness value to clamp samples at
+      maxBrightness: 100.0
+
```
Comment (callumnewlands) [inline src/main/resources/segmentation.yaml]: Any parameters with default values don't need to be specified in this file and should be removed for clarity. Similarly with segmentationRandom.yaml. Also `src/main/resources/default.yaml` needs updated to match the README (include new segmentation parameters)

**Review by callumnewlands — CHANGES_REQUESTED** [review]
The code itself looks good, but needs a few changes r.e. naming/documentation consistency/clarity

File: README.md (hunk 1 above)
Comment (DJayalath) [reply README.md]: That's a fair comment. I've renamed it to randomDataCollection - good suggestion!

File: README.md (hunk 2 above)
Comment (DJayalath) [reply README.md]: Comment added. I've also renamed endMax to maxFrames. I think this is clearer.

File: README.md (hunk 3 above)
Comment (DJayalath) [reply README.md]: Good catch! Fixed.

File: src/main/java/App.java (hunk 4 above)
Comment (DJayalath) [reply src/main/java/App.java]: Thanks for the suggestion. I've extracted this into updateRandomPosition().

File: src/main/java/segmentation/Colour.java (hunk 5 above)
Comment (DJayalath) [reply src/main/java/segmentation/Colour.java:6]: Added.

File: src/main/resources/segmentation.yaml (hunk 6 above)
Comment (DJayalath) [reply src/main/resources/segmentation.yaml]: Default values removed from `segmentation.yaml` and `segmentationRandom.yaml`. I've also now updated `default.yaml` to match the README.

**Review by callumnewlands — APPROVED** [review]

**Pull request merged**